### DISCUSSION TITLE
FR: Module migrations

### DISCUSSION THREAD

**Discussion by nilsenpaul** [craftcms/cms#8520] [Ideas] [answered]
### Description
Migrations I created for my module (in `/modules/modulehandle/src/migrations`), are not recognized, even though I set Yii's `migrationNamespaces` option like this:
   
```
controllerMap' => [
  'migrate' => [
    'class' => craft\console\controllers\MigrateController::class,
      'migrationNamespaces' => [
        'modules\modulehandle\migrations',
      ],
    ],
],
```

This might be due to this line here (thanks, @angrybrad):
https://github.com/craftcms/cms/blob/develop/src/console/controllers/MigrateController.php#L85

I have now solved this by creating the migration as a content migration in /migrations, but I think it would be nice to have all module files in one place, including migrations?

### Steps to reproduce

1. Create a module, add a migrations folder and a migration file to its src folder
2. Run ./craft migrate from Craft's root folder

**Comment by brandonkelly** [comment]
> but I think it would be nice to have all module files in one place, including migrations?

The counter argument is, it’s nice to have all project-specific migrations in one place.

The only reason Craft and plugin migrations all require their own migration tracks is because they are distributed separately. Modules are all project-specific though, so there’s no technical reason why they should get their own migration tracks.

**Comment by nilsenpaul** [comment]
> The counter argument is, it’s nice to have all project-specific migrations in one place.

Yup, that thought crossed my mind. I could live with having one project migration folder. The only thing then, that slightly bothers me, is the fact that migrations created with `./craft migrate/create` are called content migrations. Technically, I'm not migrating content. And yes, I know there's no real difference between a content migration or any other migration.

Maybe it would be enough to just have clear documentation for modules, so people don't *expect* module migrations to work ;)

For now I'll just use content migrations to migrate my module's tables. Thanks for taking the time to respond, in what must be a busy few weeks.

**Comment by brandonkelly** [comment]
> The only thing then, that slightly bothers me, is the fact that migrations created with `./craft migrate/create` are called content migrations.

I actually agree with you. We should probably rename it to “Project Migrations”, or just “Migrations”.

> Maybe it would be enough to just have clear documentation for modules, so people don't expect module migrations to work ;)

Yep we do intend to document modules. Just no time for it before 4/4.

**Comment by Mosnar** [comment]
Now that Craft 3 has matured quite a bit more, I'd like to revisit this request, as we've seen a lot more demand for "organization-specific" modules in addition to "project-specific" modules. For example, internal CRM integrations that are not intended to be released publicly, but are still used on multiple sites.

I can name at least 3-4 internal plugins & modules that I have now included on multiple websites. Additionally, I've heard a lot of talk of "internal plugins" that are submitted to the plugin store simply to claim the handle, but with no real intention to publish. The "quick fix" I think is to allow module migrations. The long term fix would be to allow organization scoped plugins of some sort.

**Comment by MatthiasBrodelet** [comment]
We are also running into this.
It's just nicer to contain all the migrations within the respective modules instead of creating 1 big migrations folder for the whole project. And we don't even plan on using these modules in other projects because they are so specific.
We mainly separate different modules for a cleaner folder structure to make maintenance easier.

**Comment by brandonkelly** [comment, marked as answer]
Craft 4.4.0 resolved this with support for “private plugins” (#12716).

We generally recommend using those over modules, for anything relatively complex, such as if migrations are needed.
